File: hooks/common/quality-events.mjs
// quality-events.mjs - запись событий следа проверок хуками набора.
// Спецификация - skills/1c-code-review/references/evidence-format.md; Python-реализация
// того же каталога - tools/quality_events.py (ее читает tools/evidence.py). Схема записи
// совпадает: номер последовательности резервируется lock-файлом (открытие с 'wx'),
// тело - JSON с сортировкой ключей, отступом 2 и завершающим переводом строки, запись
// временным файлом с переименованием. Экспорт предназначен hooks/evidence-writer.mjs и
// hooks/release-writer.mjs.

import { createHash, randomUUID } from 'node:crypto';
import { mkdir, open, rm, writeFile, readdir, stat } from 'node:fs/promises';
import { execFile as execFileCb } from 'node:child_process';
import { rename, constants } from 'node:fs';
import { promisify } from 'node:util';
import { dirname, join } from 'node:path';

const execFile = promisify(execFileCb);
const renameP = promisify(rename);

// Идентификатор сессии: буква-цифра-подчеркивание-точка-дефис, без разделителей пути.
export const SESSION_RE = /^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$/;

export class EventsError extends Error {}

// Корень git-репозитория: каталог следа один на репозиторий (tools/quality_events.py,
// repo_top). Отказ git и не-репозиторий - EventsError.
export async function repoTop(cwd) {
  try {
    const { stdout } = await execFile('git', ['rev-parse', '--show-toplevel'],
                                      { cwd, maxBuffer: 16 * 1024 * 1024 });
    return stdout.toString('utf8').trim();
  } catch (err) {
    const stderr = Buffer.isBuffer(err.stderr) ? err.stderr.toString('utf8') : String(err.stderr || '');
    throw new EventsError(`каталог не является git-репозиторием: ${cwd}${stderr ? `: ${stderr.trim()}` : ''}`);
  }
}

// Каталог событий сессии. Недопустимый идентификатор - EventsError.
export function eventsDir(top, session) {
  if (!SESSION_RE.test(session || '')) {
    throw new EventsError(`недопустимый идентификатор сессии: ${session}`);
  }
  return join(top, '.claude', '.state', 'quality', session, 'events');
}

// Рекурсивно отсортировать ключи объектов: сериализация совпадает с
// json.dumps(sort_keys=True) из tools/quality_events.py (ключи событий - ASCII).
export function sortKeysDeep(value) {
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  if (value && typeof value === 'object') {
    const out = {};
    for (const key of Object.keys(value).sort()) out[key] = sortKeysDeep(value[key]);
    return out;
  }
  return value;
}

// Момент в ISO 8601 с зоной и миллисекундами - локальное время со смещением зоны, как
// datetime.now().astimezone().isoformat() в tools/quality_events.py. Компоненты
// локального времени идут и в имя файла события; UTC отсортировался бы раньше
// Python-событий той же сессии.
export function formatIso(d) {
  const offsetMin = -d.getTimezoneOffset();
  const sign = offsetMin >= 0 ? '+' : '-';
  const abs = Math.abs(offsetMin);
  const pad = (n, w = 2) => String(n).padStart(w, '0');
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`
    + `T${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}.${pad(d.getMilliseconds(), 3)}`
    + `${sign}${pad(Math.floor(abs / 60))}:${pad(abs % 60)}`;
}

// Текущее время в том же формате (поле at события).
export function nowIso() {
  return formatIso(new Date());
}

// Часть имени файла из времени события: YYYY-MM-DDTHHMMSS-<мс> (как
// tools/quality_events.py, f"{stamp:%Y-%m-%dT%H%M%S}-{мс}"), самолексикографична.
function timePart(iso) {
  const m = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:[.,](\d{1,6}))?/.exec(iso);
  if (!m) throw new EventsError(`время события не ISO 8601: ${iso}`);
  const ms = (m[7] || '0').padEnd(3, '0').slice(0, 3);
  return `${m[1]}-${m[2]}-${m[3]}T${m[4]}${m[5]}${m[6]}-${ms}`;
}

// Занять номер последовательности отметки времени созданием lock-файла (O_CREAT|O_EXCL).
// Занятый номер перебирается дальше; lock-файлы остаются как занятые номера.
async function acquireSeq(sessionDir, stamp) {
  const { O_CREAT, O_EXCL, O_WRONLY } = constants;
  for (let seq = 0; ; seq++) {
    try {
      const handle = await open(join(sessionDir, `${stamp}-${String(seq).padStart(6, '0')}.lock`),
                               O_CREAT | O_EXCL | O_WRONLY, 0o644);
      await handle.close();
      return seq;
    } catch (err) {
      if (err.code !== 'EEXIST') throw new EventsError(`lock последовательности: ${err.message}`);
    }
  }
}

// Записать событие атомарно и вернуть путь файла. Каталог создается при отсутствии.
export async function writeEvent(top, session, event) {
  const targetDir = eventsDir(top, session);
  const sessionDir = dirname(targetDir);
  await mkdir(targetDir, { recursive: true });
  const stamp = timePart(String(event.at || ''));
  const producer = String(event.producer || 'hook');
  if (!/^[a-z]+$/.test(producer)) throw new EventsError(`недопустимый producer: ${producer}`);
  const seq = await acquireSeq(sessionDir, stamp);
  const payload = JSON.stringify(sortKeysDeep(event), null, 2) + '\n';
  // Занятое имя (коллизия случайного id) перегенерируется с тем же номером lock.
  while (true) {
    const id = randomUUID().replace(/-/g, '').slice(0, 6);
    const final = join(targetDir, `${stamp}-${String(seq).padStart(6, '0')}-${producer}-${id}.json`);
    try {
      const tmp = join(targetDir, `.tmp-${randomUUID().replace(/-/g, '')}`);
      await writeFile(tmp, payload, 'utf8');
      await renameP(tmp, final);
      return final;
    } catch (err) {
      if (err.code !== 'EEXIST') throw err;
    }
  }
}

// Список файлов каталога (без .lock и временных); отсутствующий каталог - пустой список.
export async function listEventFiles(top, session) {
  let dir;
  try {
    dir = eventsDir(top, session);
    const names = await readdir(dir);
    return names.filter((n) => n.endsWith('.json')).sort();
  } catch (err) {
    if (err.code === 'ENOENT') return [];
    throw err;
  }
}

// Вычистить каталоги сессий старше ttlMs (по времени изменения каталога сессии; его
// обновляет каждое событие - lock-файлы создаются именно там). Возвращает число удаленных.
export async function sweepStaleSessions(top, ttlMs) {
  const root = join(top, '.claude', '.state', 'quality');
  let names;
  try {
    names = await readdir(root);
  } catch (err) {
    if (err.code === 'ENOENT') return 0;
    throw err;
  }
  let removed = 0;
  for (const name of names) {
    const sessionDir = join(root, name);
    try {
      const info = await stat(sessionDir);
      if (info.isDirectory() && Date.now() - info.mtimeMs > ttlMs) {
        await rm(sessionDir, { recursive: true, force: true });
        removed++;
      }
    } catch {
      // каталог чужой или уже удален параллельным хуком - пропустить
    }
  }
  return removed;
}

// sha256 строки в hex (хеши inputHash/responseHash).
export function sha256Hex(text) {
  return createHash('sha256').update(String(text), 'utf8').digest('hex');
}

File: tools/quality_events.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Общий модуль событий следа проверок: запись, чтение, отбор прогона.

Каталог событий - <база>/<ключ>/<session>/events/ вне репозитория
(спецификация - skills/1c-code-review/references/evidence-format.md). База -
QUALITY_STATE_DIR, если переменная не пустая, иначе
<домашний каталог>/.claude/state/quality. Событие - один неизменяемый JSON-файл
с именем <время>-<номер>-<источник>-<id>.json; запись идет через временный файл
и переименование. Поврежденный JSON при чтении не поднимает исключение: файл
возвращается записью type=corrupt с именем и текстом ошибки.

Пишут события: tools/change_profile.py (scope, producer profile), tools/evidence.py
(skipped, not_verified, probe; producer cli) и хуки (applied, failed, release,
baseline, armed; producer hook) - хуки этим модулем не реализуются, для него они
только записи каталога.
"""

from __future__ import annotations

import hashlib
import json
import os
import re
import subprocess
import sys
import uuid
from datetime import datetime
from pathlib import Path

# Идентификатор сессии: буква-цифра-подчеркивание-точка-дефис, без разделителей пути.
SESSION_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")
_DRIVE_ROOT = re.compile(r"[A-Za-z]:/")


class EventsError(Exception):
    """Отказ каталога событий: недопустимый идентификатор сессии, git недоступен."""


def now_iso() -> str:
    """Текущее время в ISO 8601 с зоной и миллисекундами (поле at события)."""
    return datetime.now().astimezone().isoformat(timespec="milliseconds")


def claude_home() -> str:
    """Домашний каталог: на win32 первым USERPROFILE, иначе HOME, затем Path.home."""
    if sys.platform == "win32":
        return os.environ.get("USERPROFILE") or os.environ.get("HOME") or str(Path.home())
    return os.environ.get("HOME") or os.environ.get("USERPROFILE") or str(Path.home())


def state_base() -> Path:
    """База следа: QUALITY_STATE_DIR, если переменная не пустая, иначе домашний каталог."""
    env = os.environ.get("QUALITY_STATE_DIR", "")
    if env:
        return Path(env)
    return Path(claude_home()) / ".claude" / "state" / "quality"


def normalize_top(path: str, platform: str) -> str:
    """Нормализация корня репозитория. platform - win32, linux или darwin."""
    text = str(path).replace("\\", "/")
    while len(text) > 1 and text.endswith("/") and _DRIVE_ROOT.fullmatch(text) is None:
        text = text[:-1]
    if platform == "win32":
        text = text.lower()
    return text


def repo_key(normalized: str) -> str:
    """Ключ репозитория: сегмент пути и 12 hex sha256 нормализованной строки."""
    segment = str(normalized).rsplit("/", 1)[-1]
    segment = re.sub(r"[^A-Za-z0-9._-]", "_", segment)
    segment = re.sub(r"_+", "_", segment)
    segment = segment.strip("_.-")[:32].rstrip("_.-")
    if not segment:
        segment = "repo"
    digest = hashlib.sha256(str(normalized).encode("utf-8")).hexdigest()[:12]
    return f"{segment}-{digest}"


def _canonical_top(top: str) -> str:
    """Канонический путь. Отказ realpath - путь без изменений."""
    try:
        return os.path.realpath(top)
    except OSError:
        return top


def state_root(top: Path | str) -> Path:
    """Корень следа репозитория: <база>/<ключ>."""
    canonical = _canonical_top(os.fspath(top))
    return state_base() / repo_key(normalize_top(canonical, sys.platform))


def repo_top(repo_dir: Path | str) -> Path:
    """Корень git-репозитория по git rev-parse --show-toplevel.

    Каталог следа один на репозиторий независимо от того, каким подкаталогом задан --repo.
    """
    try:
        proc = subprocess.run(["git", "rev-parse", "--show-toplevel"],
                              cwd=str(repo_dir), capture_output=True)
    except OSError as exc:
        raise EventsError(f"git недоступен: {exc}") from exc
    if proc.returncode != 0:
        raise EventsError(f"каталог не является git-репозиторием: {repo_dir}")
    return Path(proc.stdout.decode("utf-8", errors="replace").strip())


def validate_session(session: str) -> None:
    """Проверить идентификатор сессии; недопустимый - EventsError."""
    if not SESSION_RE.match(session or ""):
        raise EventsError(f"недопустимый идентификатор сессии: {session!r}")


def session_dir(repo_dir: Path | str, session: str) -> Path:
    """Каталог сессии: <база>/<ключ>/<session>."""
    validate_session(session)
    return state_root(repo_top(repo_dir)) / session


def events_dir(repo_dir: Path | str, session: str) -> Path:
    """Каталог событий сессии вне репозитория."""
    return session_dir(repo_dir, session) / "events"


def _acquire_seq(session_dir: Path, time_part: str) -> int:
    """Занять номер последовательности отметки времени созданием lock-файла.

    Номер упорядочивает события одной миллисекунды по записи, а не по producer.
    Файл <время>-<номер>.lock в каталоге сессии создается открытием с
    O_CREAT|O_EXCL: занятый номер поднимает FileExistsError, перебор идет со
    следующего. Lock-файлы остаются в каталоге как занятые номера. Node-писатель
    событий повторяет ту же схему (evidence-format.md).
    """
    seq = 0
    while True:
        try:
            fd = os.open(str(session_dir / f"{time_part}-{seq:06d}.lock"),
                         os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o644)
        except FileExistsError:
            seq += 1
            continue
        os.close(fd)
        return seq


def _filename(event: dict, session_dir: Path) -> str:
    """Имя файла события: время, номер последовательности, producer и случайный id.

    Время формата YYYY-MM-DDTHHMMSS-<мс> самолексикографично; номер - 6 цифр
    монотонной последовательности отметки времени, поэтому сортировка имен
    совпадает с порядком записи и в пределах одной миллисекунды. Поле at обязано
    быть ISO 8601; иначе EventsError.
    """
    try:
        stamp = datetime.fromisoformat(event["at"])
    except (KeyError, TypeError, ValueError) as exc:
        raise EventsError("событие без поля at или с не-ISO временем") from exc
    producer = event.get("producer", "cli")
    if not re.fullmatch(r"[a-z]+", str(producer)):
        raise EventsError(f"недопустимый producer: {producer!r}")
    time_part = f"{stamp:%Y-%m-%dT%H%M%S}-{stamp.microsecond // 1000:03d}"
    seq = _acquire_seq(session_dir, time_part)
    return f"{time_part}-{seq:06d}-{producer}-{uuid.uuid4().hex[:6]}.json"


def write_event(repo_dir: Path | str, session: str, event: dict) -> Path:
    """Записать событие атомарно и вернуть путь файла.

    Тело - UTF-8 JSON с сортировкой ключей, отступом 2 и завершающим переводом
    строки. Временный файл пишется в каталог событий и переименовывается; занятое
    имя перегенерируется (id случайный, коллизия маловероятна, но проверяется).
    """
    target_dir = events_dir(repo_dir, session)
    session_dir = target_dir.parent
    target_dir.mkdir(parents=True, exist_ok=True)
    payload = json.dumps(event, ensure_ascii=False, indent=2, sort_keys=True) + "\n"
    final = target_dir / _filename(event, session_dir)
    while final.exists():
        final = target_dir / _filename(event, session_dir)
    tmp = target_dir / f".tmp-{uuid.uuid4().hex}"
    tmp.write_text(payload, encoding="utf-8", newline="\n")
    os.replace(tmp, final)
    return final


def read_events(repo_dir: Path | str, session: str) -> list[dict]:
    """Прочитать события сессии в порядке имен; поврежденные - записью type=corrupt.

    Каждое событие дополняется служебным ключом _file (имя файла): по нему ссылаются
    пропуски (ref) и строится диагностика. Отсутствующий каталог - пустой список.
    """
    directory = events_dir(repo_dir, session)
    if not directory.is_dir():
        return []
    events: list[dict] = []
    for path in sorted(directory.glob("*.json")):
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError, UnicodeDecodeError) as exc:
            events.append({"type": "corrupt", "file": path.name, "error": str(exc)})
            continue
        if not isinstance(data, dict):
            events.append({"type": "corrupt", "file": path.name,
                           "error": "корень JSON не объект"})
            continue
        data["_file"] = path.name
        events.append(data)
    return events


def select_run(events: list[dict], diff_hash: str) -> dict | None:
    """Отобрать прогон: последнее scope с diffHash и события с тем же хешем после него.

    Возвращает {"scope": событие, "events": [события после него с тем же diffHash]}
    либо None, когда scope с таким хешем в каталоге нет. Порядок берется по позициям
    в списке (он же порядок имен файлов).
    """
    run = None
    for index, event in enumerate(events):
        if event.get("type") == "scope" and event.get("diffHash") == diff_hash:
            run = {"scope": event,
                   "events": [e for e in events[index + 1:]
                              if e.get("diffHash") == diff_hash]}
    return run
